docs/Testing.md

```diff
@@ -0,0 +1,454 @@
+
+### Using the Test Runner
+
+First, add Robolectric to your `dependencies`
+(e.g., `testImplementation  "org.robolectric:robolectric:3.3.2"`), if you do not
+have it for other reasons already.
+
+Then, create a JUnit test class that uses 
+`DopplContextDelegateTestRunner` as the test runner:
+
+```java
+@RunWith(DopplContextDelegateTestRunner.class)
+public class ExampleContextTest {
+  @Test
+  public void testSomething() {
+    // TODO: a really cool test
+  }
+}
+```
+
+Here, `DopplContextDelegateTestRunner` comes from the `co.doppl.lib:androidbasetest`
+dependencies that you should have set up for the `testImplementation` and
+the `testDoppl` dependency scopes.
+
+### Getting an Application Context
+
+Then, your test methods can use `DopplRuntimeEnvironment.getApplication()`
+to get a mock `Application` singleton:
+
+```java
+@RunWith(DopplContextDelegateTestRunner.class)
+public class ExampleContextTest {
+  @Test
+  public void testSomething() {
+    final SharedPreferences prefs=
+      DopplRuntimeEnvironment
+        .getApplication()
+        .getSharedPrefences(TEST_PREFS, Context.MODE_PRIVATE);
+
+    // TODO: a really cool test involving SharedPreferences
+  }
+}
+```
+
+Doppl will arrange to give you a suitable mock, both on Android (via
+Robolectric) or on iOS.
+
+## Using Spies with Mockito
+
+Doppl integrates with J2objc's support for Mockito. However, at the present time,
+that support is limited to mocks. If you want to spy an object &mdash; perhaps
+to confirm that methods were invoked &mdash; J2objc does not support that.
+
+However, *Doppl* supports object spies, though it takes a bit more work to set up,
+beyond simply having Mockito as a test dependency (which is all that you need
+for simple mocks). This is through a `@MockGen` annotation, backed by a compile-time
+annotation processor that generates code to support object spying on iOS.
+
+Note that this is a stop-gap implementation, while improvements to J2objc's
+Mockito support are still in progress.
+
+### Writing a Test Using a Spy
+
+On the Android side, you can go ahead and create your unit test that
+uses Mockito's spy capability:
+
+```java
+@Test
+public void timer() throws InterruptedException {
+  ScheduledExecutorService scheduler=Executors.newSingleThreadScheduledExecutor();
+  Runnable thingy=spy(new Runnable() {
+    @Override
+    public void run() {
+      // TODO: something useful
+    }
+  });
+
+  verify(thingy, never()).run();
+  scheduler.schedule(thingy, 200, TimeUnit.MILLISECONDS);
+  Thread.sleep(500);
+  verify(thingy, times(1)).run();
+}
+```
+
+This should run fine in your Android unit tests... on Android. It will not
+work on iOS, until we add in `@MockGen`.
+
+### Adding the Annotation Processor
+
+The annotation processor for `@MockGen` is `co.doppl.lib:testapt`, and you
+will need to add this to your dependencies:
+
+```groovy
+testAnnotationProcessor `co.doppl.lib:testapt:0.8.5`
```

Comment:
Probably mean to use single quotes rather than backticks here?

Reply:
I'm running into strange copy-paste problems with Sublime Text 3, changing my quotes into backticks sometimes, for some reason. Might be tied to pasting into GitHub-style fenced code blocks (the triple-backtick syntax). Regardless, it's aggravating, as I don't always notice the change. Thanks!

Reply:
Fixed in update.